### PR TITLE
feat(types): add vue-template-compiler types

### PR DESCRIPTION
**Pull request by ktsn**
**What kind of change does this PR introduce?** (check at least one)

- [ ] Bugfix
- [x] Feature
- [ ] Code style update
- [ ] Refactor
- [ ] Build-related changes
- [ ] Other, please describe:

**Does this PR introduce a breaking change?** (check one)

- [ ] Yes
- [x] No

If yes, please describe the impact and migration path for existing applications:

**The PR fulfills these requirements:**

- [x] It's submitted to the `dev` branch for v2.x (or to a previous version branch), _not_ the `master` branch
- [ ] When resolving a specific issue, it's referenced in the PR's title (e.g. `fix #xxx[,#xxx]`, where "xxx" is the issue number)
- [x] All tests are passing: https://github.com/vuejs/vue/blob/dev/.github/CONTRIBUTING.md#development-setup
- [ ] New/updated tests are included

If adding a **new feature**, the PR's description includes:
- [x] A convincing reason for adding this feature (to avoid wasting your time, it's best to open a suggestion issue first and wait for approval before working on it)

**Other information:**
I just remembered I wrote `vue-template-compiler` typings for my own project when seeing [`component-compiler-utils` types](https://github.com/vuejs/component-compiler-utils/blob/master/lib/types.ts) 😄 
It would be convenient for anyone who writes static analysis / customized compiler / build tools in TypeScript.

### CODE REVIEW

File: packages/vue-template-compiler/types/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+  ssrOptimizability?: number;
+}
+
+export interface ASTText {
+  type: 3;
+  text: string;
+  static?: boolean;
+  isComment?: boolean;
+  // 2.4 ssr optimization
+  ssrOptimizability?: number;
```
Comment (HerringtonDarkholme) [inline packages/vue-template-compiler/types/index.d.ts]: Is it better to add specific literal type to model optimization? IIRC ssrOptimizability has a enum like constants.

File: packages/vue-template-compiler/types/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+interface CompiledResult {
+  ast: ASTElement | undefined;
+  render: string;
+  staticRenderFns: string[];
+  errors: string[];
+  tips: string[];
+}
+
+interface CompiledResultFunctions {
+  render: Function;
```
Comment (HerringtonDarkholme) [inline packages/vue-template-compiler/types/index.d.ts]: We can provide a more precise type: `() => VNode`

**Comment by ktsn** [comment]
@HerringtonDarkholme Fixed 🙂

File: packages/vue-template-compiler/types/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+interface CompiledResult {
+  ast: ASTElement | undefined;
+  render: string;
+  staticRenderFns: string[];
+  errors: string[];
+  tips: string[];
+}
+
+interface CompiledResultFunctions {
+  render: (this: Vue) => VNode;
```
Comment (HerringtonDarkholme) [inline packages/vue-template-compiler/types/index.d.ts]: I don't know whether `this` annotation will work. I have encountered problem before due to Vue is an invariant generic, which causes assignment problems in strict variance mode. A test case might be good. Or leaving this annotation is also acceptable to me.

Other than this, this pull request looks good to me!

Comment (ktsn) [reply packages/vue-template-compiler/types/index.d.ts]: I agree. I just removed the this type annotation.

**Comment by wadetandy** [comment]
@HerringtonDarkholme @blake-newman Any chance we can get this merged soon? Would be a huge help!

**Comment by ypresto** [comment]
I'm depending on type def of complier found in `@vue/component-compiler-utils`. It has bad cohesion. Hoping this to be merged.

**Pull request merged**